drivers/sensor/bma4xx/bma4xx.c

```diff
@@ -0,0 +1,687 @@
+{
+	struct bma4xx_data *bma4xx = dev->data;
+
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+	const enum sensor_channel *const channels = cfg->channels;
+	const size_t num_channels = cfg->count;
+
+	uint32_t min_buf_len = sizeof(struct bma4xx_encoded_data);
+	struct bma4xx_encoded_data *edata;
+	uint8_t *buf;
+	uint32_t buf_len;
+	int rc;
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	/* Prepare response */
+	edata = (struct bma4xx_encoded_data *)buf;
+	edata->header.is_fifo = false;
+	edata->header.accel_fs = bma4xx->range;
+	edata->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+	edata->has_accel = 0;
+	edata->has_temp = 0;
+
+	/* Determine what channels we need to fetch */
+	for (int i = 0; i < num_channels; i++) {
+		switch (channels[i]) {
+		case SENSOR_CHAN_ACCEL_X:
+		case SENSOR_CHAN_ACCEL_Y:
+		case SENSOR_CHAN_ACCEL_Z:
+		case SENSOR_CHAN_ACCEL_XYZ:
+			edata->has_accel = 1;
+			break;
+#ifdef CONFIG_BMA4XX_TEMPERATURE
+		case SENSOR_CHAN_DIE_TEMP:
+			edata->has_temp = 1;
+			break;
+#endif /* CONFIG_BMA4XX_TEMPERATURE */
+		default:
+			LOG_ERR("Requested unsupported channel ID %d", channels[i]);
+			return -ENOTSUP;
+		}
+	}
+
+	if (edata->has_accel) {
+		rc = bma4xx_sample_fetch(dev, &edata->accel_xyz[0], &edata->accel_xyz[1],
+					 &edata->accel_xyz[2]);
+		if (rc != 0) {
+			LOG_ERR("Failed to fetch accel samples");
+			rtio_iodev_sqe_err(iodev_sqe, rc);
+			return rc;
+		}
+	}
+
+#ifdef CONFIG_BMA4XX_TEMPERATURE
+	if (edata->has_temp) {
+		rc = bma4xx_temp_fetch(dev, &edata->temp);
+		if (rc != 0) {
+			LOG_ERR("Failed to fetch temp sample");
+			rtio_iodev_sqe_err(iodev_sqe, rc);
+			return rc;
+		}
+	}
+#endif /* CONFIG_BMA4XX_TEMPERATURE */
+
+	rtio_iodev_sqe_ok(iodev_sqe, 0);
+
+	return 0;
+}
+
+int bma4xx_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *cfg = iodev_sqe->sqe.iodev->data;
+
+	if (!cfg->is_streaming) {
+		return bma4xx_submit_one_shot(dev, iodev_sqe);
+	}
+	/* TODO: Add streaming support */
+
+	return -ENOTSUP;
+}
+
+/*
+ * RTIO decoder
+ */
+
+static int bma4xx_decoder_get_frame_count(const uint8_t *buffer, enum sensor_channel channel,
+					  size_t channel_idx, uint16_t *frame_count)
+{
+	const struct bma4xx_encoded_data *edata = (const struct bma4xx_encoded_data *)buffer;
+	const struct bma4xx_decoder_header *header = &edata->header;
+
+	if (channel_idx != 0) {
+		return -ENOTSUP;
+	}
+
+	if (!header->is_fifo) {
+		switch (channel) {
+		case SENSOR_CHAN_ACCEL_X:
+		case SENSOR_CHAN_ACCEL_Y:
+		case SENSOR_CHAN_ACCEL_Z:
+		case SENSOR_CHAN_ACCEL_XYZ:
+		case SENSOR_CHAN_DIE_TEMP:
+			*frame_count = 1;
```

Comment:
I think you should check `edata->has_accel` and `edata->has_temp` here

Reply:
Set `frame_count` to `0` if accel or temp data is not present? Should it return `0` or `-ENODATA`?